### PR TITLE
feat: reject misplaced bump bullets in RELEASE.md (DT-145 follow-up)

### PR DESCRIPTION
**Pull request by lucasmundim**
## Summary

Two new validation rules in `ChangelogParser` to catch mis-edits the cascade `bump-strategy.sh` could previously produce. Pairs with the producer-side fix in [PlainsightAI/openfilter#104](https://github.com/PlainsightAI/openfilter/pull/104) — that PR stops the script from creating new bad files; this PR fails CI on any RELEASE.md (including hand-written) that has the same shape.

The parser is now driven by `marked.lexer` (CommonMark AST) rather than regex/string-splits, so bump-shaped text inside fenced code blocks or HTML comments doesn't false-positive.

### Rule 1 — bump bullets in the file intro

A list token above the first `## ` header whose first line matches `- Bump <pkg> to X.Y.Z` throws with the offending bullet(s). Catches the **FaceGuard pattern**: bump bullets accumulating in the file preamble because a prior bump landed there and subsequent runs anchored off it.

The rule is **scoped to the bump-bullet shape** (not "any bullet in the intro") so projects whose intros legitimately use bullet lists — feature summaries, pointers to a canonical changelog — are not affected.

### Rule 2 — `- Bump <pkg> to X.Y.Z` under a released section

A post-parse validator scans each non-Unreleased entry's `list` tokens for bullets matching the bump pattern and throws with the section and offending bullet. The mechanical bump pattern is the cascade automation's signal — its presence inside a tagged release block is by definition misplaced.

Catches the **[filter-sweet-green-subject-data-aggregator #39](https://github.com/PlainsightAI/filter-sweet-green-subject-data-aggregator/pull/39/files)** pattern, where a bump bullet landed inside `## v0.1.27 - 2026-04-24`.

Both errors surface as `core.setFailed(error.message)` from `main.ts` with actionable text naming the offending section and bullet.

### Implementation notes

- `marked` is added as a runtime dep and bundled into `build/dist/index.js` via `ncc` — zero install cost for action consumers (bundle grew ~44kB).
- The intro and released-section rules walk **top-level** `list` tokens only; the cascade `bump-strategy.sh` always emits flat top-level bullets, so this matches the actual threat model. A nested-list bump bullet would slip past — deliberately out of scope for the current rule set.
- `bumpBulletRegex` matches `-` markers only (cascade always emits `-`). A hand-typed `*` or `+` bump-shaped bullet under a released section would not trip the rule — also by design; this rule exists to catch the mechanical output.
- Description is now rebuilt from `token.raw` (preserves per-line interior whitespace) rather than per-line `.trim()`. A regression test pins the current shape — `openfilter/.github/workflows/create-release.yaml` consumes `outputs.description` directly, so future parser swaps must preserve this.
- `jest.config.js` routes `marked` through its UMD build via `moduleNameMapper` because marked v18's default resolution is ESM and `ts-jest` runs in CJS mode. ncc production bundling is unaffected.

## Tests

62 Jest tests pass (51 pre-existing + 11 added in this PR).

**Intro rule (Rule 1)**
- throws on the FaceGuard pattern (bump bullet above `## [Unreleased]`)
- reports multiple misplaced bump bullets in one error
- allows plain-paragraph intros (Keep a Changelog convention)
- allows non-bump bullet lists in the intro (e.g. feature summaries)

**Released-section rule (Rule 2)**
- throws on the SGSDA #39 pattern (named section + bullet in error)
- throws on bump bullet in released section even without a date
- allows bump bullets correctly placed under `[Unreleased]`
- does NOT fire on narrative `Bumped ...` prose in release bodies
- accepts variant package names (`filter-faceblur` etc.) in the rule

**Behavior pins**
- description preserves per-line interior whitespace post-marked refactor

Coverage on `ChangelogParser.ts`: 100% statement / 100% branch.

## Notes

- DT-145: https://plainsight-ai.atlassian.net/browse/DT-145
- Companion PR (producer fix): PlainsightAI/openfilter#104

## Test plan

- [ ] Existing consumers of this action (openfilter `check-release-log`, any other repos) continue to parse their valid RELEASE.md without new failures
- [ ] On a synthetic PR that adds a bump bullet under a released section, this action fails with the new error message

### CODE REVIEW

File: src/ChangelogParser.ts [hunk 1]
```diff
@@ -3,7 +3,18 @@ import { Changelog, ChangelogEntry } from "./Changelog";
...
+  private static readonly bulletLineRegex = /^[-*+]\s/;
+
+  // `- Bump <pkg> to <semver>` — the mechanical bullet the cascade
+  // bump-strategy automation produces. Must only appear under
+  // `[Unreleased]`; finding it inside a tagged release header means the
+  // bump landed in the wrong block.
+  private static readonly bumpBulletRegex = /^-\s+Bump\s+\S+\s+to\s+\d+(?:\.\d+)+/;
+
   static parseChangelog(changelog: string): Changelog {
     const changelogNoIntro = ChangelogParser.removeChangelogIntro(changelog);
```
Comment (stwilt) [inline src/ChangelogParser.ts]: While the current string-split and regex logic is fully functional and passes all tests, we can completely eliminate regex-based layout parsing (and the risk of false positives if someone mentions a `- Bump` bullet inside an inline/fenced code block or comment) by using a proper AST-based approach.

Since this Action is compiled into a single self-contained bundle via `@vercel/ncc` inside `build/dist/`, adding `marked` adds absolutely zero runtime overhead to the action's consumers.

Here is a complete, drop-in replacement for `ChangelogParser.ts` using `marked` that is robust, context-aware, and behaves identically with all existing test cases:

```typescript
import { marked } from "marked";
import { Changelog, ChangelogEntry } from "./Changelog";

export class ChangelogParser {
  static parseChangelog(changelog: string): Changelog {
    const tokens = marked.lexer(changelog);
    const entries: ChangelogEntry[] = [];
    let currentEntry: {
      version: string;
      status: string;
      date?: string;
      descriptionTokens: marked.Token[];
    } | null = null;

    const introTokens: marked.Token[] = [];

    for (const token of tokens) {
      if (token.type === "heading" && token.depth === 2) {
        const header = token.text;
        const unreleasedMatch = header.match(/^\[\s*unreleased\s*\]$/i);
        if (unreleasedMatch) {
          currentEntry = {
            version: "unreleased",
            status: "unreleased",
            date: undefined,
            descriptionTokens: [],
          };
          entries.push(currentEntry as unknown as ChangelogEntry);
        } else {
          const headerMatch = header.match(
            /^v([0-9]+)\.([0-9]+)\.([0-9]+)(?:\.([0-9]+)-((?:dev|rc|int)))?(?:\s*-\s*(\d{4}-\d{2}-\d{2}))?$/
          );
          if (headerMatch) {
            const [, major, minor, patch, build, suffix, date] = headerMatch;
            const isProd = !suffix;
            const version = isProd
              ? `v${major}.${minor}.${patch}`
              : `v${major}.${minor}.${patch}.${build}-${suffix}`;

            currentEntry = {
              version,
              status: isProd ? "release" : "prerelease",
              date,
              descriptionTokens: [],
            };
            
            const entry: ChangelogEntry = {
              version,
              versionMajor: major,
              versionMinor: minor,
              versionPatch: patch,
              buildNumber: build,
              suffix,
              status: isProd ? "release" : "prerelease",
              date,
              description: "",
            };
            (entry as any).descriptionTokens = currentEntry.descriptionTokens;
            entries.push(entry);
          } else {
            throw new Error("Could not parse CHANGELOG entry:\n" + header);
          }
        }
      } else {
        if (currentEntry === null) {
          introTokens.push(token);
        } else {
          currentEntry.descriptionTokens.push(token);
        }
      }
    }

    // Validate intro contains no bullets (lists)
    ChangelogParser.validateIntroTokensHaveNoLists(introTokens);

    // Build description strings from tokens and validate bump bullets
    for (const entry of entries) {
      const descTokens = (entry as any).descriptionTokens || [];
      entry.description = descTokens.map((t: marked.Token) => t.raw).join("").trim();
      delete (entry as any).descriptionTokens;

      if (entry.status !== "unreleased") {
        ChangelogParser.validateNoBumpBulletsInTokens(descTokens, entry.version);
      }
    }

    ChangelogParser.validateUniqueVersions(entries);
    return new Changelog(entries);
  }

  private static validateIntroTokensHaveNoLists(tokens: marked.Token[]): void {
    const offending: string[] = [];
    const traverse = (t: marked.Token) => {
      if (t.type === "list") {
        for (const item of (t as marked.Tokens.List).items) {
          offending.push(item.raw.trim());
        }
      }
    };
    tokens.forEach(traverse);

    if (offending.length > 0) {
      throw new Error(
        `Changelog intro (above the first "## " header) contains bullet ` +
        `line(s) that look like misplaced changelog entries: ` +
        offending.map(b => `"${b}"`).join(", ") +
        `. Move them under "## [Unreleased]".`
      );
    } 
  }

  private static validateNoBumpBulletsInTokens(
    tokens: marked.Token[],
    version: string
  ): void {
    const bumpBulletRegex = /^-\s+Bump\s+\S+\s+to\s+\d+(?:\.\d+)+/;
    const offending: string[] = [];

    const traverse = (t: marked.Token) => {
      if (t.type === "list") {
        for (const item of (t as marked.Tokens.List).items) {
          if (bumpBulletRegex.test(item.raw.trim())) {
            offending.push(item.raw.trim());
          }
        }
      }
    };
    tokens.forEach(traverse);

    if (offending.length > 0) {
      throw new Error(
        `Bump bullet found in released section "${version}": ` +
        `"${offending[0]}". Bump bullets must live under ` +
        `"## [Unreleased]"; the cascade bump-strategy script writes there.`
      );
    }
  }
}
```

This structural AST approach isolates actual list tokens, cleanly resolving the exact concerns raised in your related `openfilter` PR.

**Review by stwilt — COMMENTED** [review]
I have completed a thorough review of the PR implementing validation rules for misplaced bump bullets (as a follow-up to DT-145). The current implementation is exceptionally clean, robust, and backed by a comprehensive test suite covering all edge cases. 

However, to make the parser completely bulletproof, future-proof, and to avoid brittle regex/string-based layout parsing (such as splitting on `\n## ` or scanning lines blindly which can cause false positives inside code blocks), I strongly recommend transitioning to a proper CommonMark AST-based approach using a lightweight, zero-dependency Markdown parser like `marked`. 

I have drafted and proposed a fully compatible implementation using `marked` in the inline comments below. Excellent work!

**Review by lucasmundim — COMMENTED** [review]
Thanks for the review! Pushed `ef6e8ef` adopting the marked suggestion.

- `ChangelogParser.ts` is now driven by `marked.lexer`. Entry boundaries come from H2 (`heading` depth=2) tokens; the version-format regex is kept verbatim for parsing the heading text. Link reference defs are filtered by token type, matching the previous `removeLinkLabels` behavior.
- `validateNoListInIntro` and `validateNoBumpBulletsInRelease` walk only `list` tokens — so a `- Bump openfilter to X.Y.Z` line inside a fenced code block or HTML comment no longer false-positives. That was the most concrete correctness win from this swap.
- `jest.config.js`: marked v18 ships ESM-only, so Jest gets routed through the UMD build via `moduleNameMapper`. ncc production bundling is unaffected — same CJS single-file output as before, just ~44kB larger.
- All 61 tests still pass (51 pre-existing + 10 added in this PR). Coverage on `ChangelogParser.ts`: 100% statement / 100% branch.

File: src/ChangelogParser.ts (hunk 1 above)
Comment (lucasmundim) [reply src/ChangelogParser.ts]: Done in `ef6e8ef` — `ChangelogParser` is now driven by `marked.lexer`. Entry boundaries come from H2 tokens, the existing version-format regex parses the heading text, and link reference defs are filtered by token type (preserves the old `removeLinkLabels` behavior). The two new validation rules walk only `list` tokens, so a `- Bump ...` line in a fenced code block or HTML comment in release notes no longer trips them — that's the concrete correctness win you flagged.

Small structural deltas from your draft: dropped the `(entry as any) .descriptionTokens` hack by tracking the token list in a `PendingEntry` shape that's separate from `ChangelogEntry`, and the intro rule reports every offending bullet (not just the first) to match the existing preamble-error format. ncc bundling is unchanged — marked ends up in `build/dist/index.js` and adds ~44kB. Jest needed `moduleNameMapper` since marked v18 is ESM-only, but that's test-config only, not runtime.

**Review by shingonoide — COMMENTED** [review]
Nice follow-up to DT-145. Two thoughtfully scoped validation rules plus the AST refactor stwilt suggested — CI is green, coverage stays at 100% on `ChangelogParser.ts`, and the new tests cover both the FaceGuard intro pattern and the SGSDA #39 in-released-section pattern. The marked-driven walk is genuinely the right move: a `- Bump ...` line inside a fenced block inside a release body will no longer false-positive.

A few non-blocking observations, mostly for awareness:

1. **PR description is now stale relative to HEAD.** The "Notes" section still talks about `removeChangelogIntro` and the `index >= 0` adjacent fix, but after `ef6e8ef` that helper no longer exists — the parser is fully AST-driven now. Worth a quick body refresh before merge so the merge-commit history matches what actually landed (the individual commit messages themselves are accurate).

2. **Validators only walk top-level tokens.** In `src/ChangelogParser.ts`, both `validateNoListInIntro` (`:124-144`) and `validateNoBumpBulletsInRelease` (`:150-167`) iterate the entry's top-level token array and only inspect `t.type === "list"` at that level. A nested-list bump bullet (e.g. `- Changed\n  - Bump openfilter to 1.1.0` under a released H2) lives in `item.tokens` and would not be caught. The cascade `bump-strategy.sh` only emits flat top-level bullets, so this matches the actual threat model — but if you ever want to harden it further, a small recursive walk over `item.tokens` would close the gap without altering current behavior.

3. **`validateNoListInIntro` false-positive risk for legitimate intro bullets.** The validator fires on *any* list token before the first `## ` header, with an error message saying the bullets "look like misplaced changelog entries." A project whose intro legitimately uses a bullet list (e.g. a features summary, a one-line bullet pointer to the canonical changelog) will now fail. There's no opt-out in `action.yml`. Probably fine for the current downstream set, but worth checking each consumer's RELEASE.md before merge — and if any have non-bump intro bullets, either rewrite their intros or add an input flag to scope the validator to bump-shaped bullets only.

4. **Subtle `description` whitespace shift propagates downstream.** The previous parser did `lines.map(line => line.trim()).join("\n")` per entry; the new one does `descriptionTokens.map(t => t.raw).join("").trim()`. Net effect: per-line interior whitespace is now preserved verbatim (only the full description is trimmed). `openfilter/.github/workflows/create-release.yaml` consumes `outputs.description` directly as GitHub release notes content, so the shift will show up there. Realistic RELEASE.md prose has no meaningful per-line whitespace, so this is likely inert in practice — but it's an undocumented behavioral change with no test guarding either direction. A targeted equality test would prevent future regressions in either direction.

5. **Rule 2 only matches `-` markers, Rule 1 matches all three.** `bumpBulletRegex` keys on `^-\s+Bump\s+\S+\s+to\s+\d+(?:\.\d+)+`, so a hypothetical hand-typed `* Bump openfilter to 1.1.0` under a released section would slip past. Given the rule's purpose is catching the cascade script's mechanical output (always `-`), the asymmetry is intentional and reasonable — calling it out so future readers understand it is by design.

6. **Marked Jest workaround via `moduleNameMapper`.** Routing `marked` through the UMD build in `jest.config.js` is a tidy workaround for the v18 package whose `main`/`module` both point at the ESM build against ts-jest's CJS output. The inline comment says "ships ESM-only from v5" — small nit: marked v18 actually ships both `lib/marked.esm.js` and `lib/marked.umd.js`; what's "ESM-only" is the default resolution, not the package layout. Worth a one-word tweak so a future maintainer who bumps `marked` knows to verify the UMD path still exists in the new version.

Everything reads cleanly otherwise — tokenization choice is solid, error messages name the offending section and bullet, and the test bodies pin both the positive and negative cases for each rule.

File: src/ChangelogParser.ts [hunk 2]
```diff
@@ -1,88 +1,168 @@
...
+  // Bullet lines in the intro almost always mean a changelog entry was
+  // appended above the first version header (FaceGuard case in DT-145).
+  // Driving off `list` tokens — not raw-line `^[-*+]` matches — means a
+  // bullet-looking line inside a code block or HTML comment won't trip
+  // the rule.
+  private static validateNoListInIntro(tokens: Token[]): void {
+    const offending: string[] = [];
+    for (const t of tokens) {
+      if (t.type === "list") {
+        for (const item of (t as Tokens.List).items) {
```
Comment (stwilt) [inline src/ChangelogParser.ts]: `marked`'s `list` token covers both ordered (`1. item`) and unordered (`- item`) lists — `t.ordered` distinguishes them. The error message says "look like misplaced changelog entries", which is wrong for an intro like:

```markdown
# Changelog
1. Breaking changes require a manual release.
2. See README for deploy instructions.

## v1.2.3
```

That intro is perfectly valid but this validator throws with a misleading message about misplaced changelog entries. The rewriter only writes unordered bullets (`- Bump …`), so the validation goal only requires checking `!t.ordered`. A `t.ordered` guard here closes the false-positive without changing behaviour for any current consumer (all have prose-only intros today):

```typescript
if (t.type === "list" && !(t as Tokens.List).ordered) {
```

**Review by stwilt — COMMENTED** [review]
Cross-repo contract verified (openfilter/bump-strategy.sh → this parser) and the `[Unreleased]` anchor fix in openfilter#104 aligns correctly. Two issues worth addressing before merge.

**Contract check (bump-strategy.sh ↔ parser):** The rewriter emits `- Bump openfilter to ${OF_VERSION}` (flat, leading `- `, no indentation). `bumpBulletRegex` at `src/ChangelogParser.ts` L14 matches exactly that form. No whitespace or indentation mismatch. ✓

**`[Unreleased]` anchor (openfilter#104 vs. this parser):** `parseHeading` at `src/ChangelogParser.ts` L75 uses `/^\[\s*unreleased\s*\]$/i`, which matches `[Unreleased]` — the form openfilter#104's rewriter now writes everywhere. The old pre-fix form `(unreleased)` (parentheses) is gone from all live consumer RELEASE.md files checked. ✓

**Two net-new issues inline.**

**Review by lucasmundim — COMMENTED** [review]
Thanks @shingonoide — all the actionable items addressed in `300334c`, plus a PR-body refresh.

**Landed:**

- **#1 PR body refresh.** Updated to match HEAD — the `removeChangelogIntro` / `index >= 0` notes are gone since that helper no longer exists after the marked refactor.
- **#3 intro false-positive risk.** `validateNoListInIntro` now scopes to the bump-bullet shape (`- Bump <pkg> to X.Y.Z`) instead of "any list above the first `## `". Projects whose intros legitimately use bullet lists (feature summaries, pointers to a canonical changelog) no longer fail. The FaceGuard pattern is still caught — that case is also bump-shaped. Error message updated to "contains bump bullet(s) that look like misplaced cascade output" to reflect the narrower scope. Added a positive test (`should allow a non-bump bullet list in the intro`) to guard against re-broadening.
- **#4 description whitespace shift.** Added `should preserve per-line interior whitespace in description (post-marked behavior pin)` — `openfilter/.github/workflows/create-release.yaml` reads `outputs.description` directly as release-notes content, so locking the current shape prevents silent regressions either direction on a future parser swap.
- **#5 `-`-only matching in Rule 2.** Added a comment on `bumpBulletRegex` documenting that `-`-only is intentional (the rule exists to catch the cascade script's mechanical output, which always emits `-`; a hand-typed `* Bump ...` slipping past is by design). Same reasoning now applies to Rule 1 since it shares the regex.
- **#6 Jest config comment.** Fixed — marked v18 ships both ESM and UMD; what's "ESM-only" is the default resolution (`main`/`module` both point at `marked.esm.js`). Updated to note that, plus a reminder to verify the UMD path on future bumps.

**Deliberately skipped:**

- **#2 nested-list recursion.** The cascade `bump-strategy.sh` only emits flat top-level bullets — the script's RELEASE.md handler is line-oriented and writes a single `- Bump openfilter to X.Y.Z` directly under `### Changed`. A nested bump bullet can't be produced by the automation, so the recursive walk would be defending against a threat that can't reach us. Documented as deliberate in the implementation notes in the refreshed PR body so a future reader understands it's by design rather than oversight.

62 tests pass (was 61 — dropped the `* and + bullets in intro` test which was obsolete under the narrower rule, added two). Coverage on `ChangelogParser.ts` remains 100% / 100%.

**Review by lucasmundim — COMMENTED** [review]
Thanks @stwilt — appreciate the cross-repo contract verification (both the `bumpBulletRegex` ↔ rewriter alignment and the `[Unreleased]` regex ↔ openfilter#104 emit shape). That's the kind of cross-PR check that's easy to skip.

Pushed `7525ca6` adding the explicit `!(t as Tokens.List).ordered` skip in `validateNoListInIntro`. Worth flagging: the example you cited actually parses cleanly on HEAD — `bumpBulletRegex`'s `^-` anchor already filtered ordered-list items, so an ordered intro list never reached the offending-bullets list. I smoke-tested your exact `1. Breaking changes …` / `2. See README …` input before pushing — it returned `["v1.2.3"]`, no throw. The error message you quoted (*"look like misplaced changelog entries"*) is also pre-`300334c` — the current message is *"look like misplaced cascade output"*. You may have been looking at an earlier revision of the file.

That said, the guard is a real improvement: it makes the intent visible at the call site and future-proofs against any broadening of `bumpBulletRegex`. Added a regression test pinning that an ordered intro list passes. 63 tests pass; bundle unchanged.

**One question:** the body says *"Two net-new issues inline"* but only one inline came through (the ordered-list one). Was the second meant to land somewhere else, or did it get dropped? Happy to address it if you re-post.

File: src/ChangelogParser.ts (hunk 2 above)
Comment (lucasmundim) [reply src/ChangelogParser.ts]: Done in `7525ca6` — added the explicit `(t as Tokens.List).ordered` skip with a comment noting it's defense-in-depth on top of `bumpBulletRegex`'s `^-` anchor. The example you cited actually parses cleanly on HEAD (I smoke-tested before pushing — your `1. Breaking changes …` input returns `["v1.2.3"]`, no throw), because the regex filter rejected ordered-item raw text already. The guard makes the intent explicit so a future change that broadens the regex doesn't accidentally re-introduce false positives on ordered lists.

Also added `should allow an ordered list in the intro` as a regression test pinning that behavior. Side note on the message text: current error reads *"contains bump bullet(s) that look like misplaced cascade output"* (updated in `300334c`); the *"misplaced changelog entries"* phrasing you quoted was pre-`300334c`.

**Review by stwilt — COMMENTED** [review]
#1 (ordered-list false positive) — fixed cleanly. `7525ca6` adds the explicit `(t as Tokens.List).ordered` guard at [`src/ChangelogParser.ts` L142](https://github.com/PlainsightAI/changelog-parser-action/blob/7525ca6521d48af8c70b9b88aa4a6e6ed5eb766c/src/ChangelogParser.ts#L142), with a pinning test at [`src/ChangelogParser.test.ts` L354](https://github.com/PlainsightAI/changelog-parser-action/blob/7525ca6521d48af8c70b9b88aa4a6e6ed5eb766c/src/ChangelogParser.test.ts#L354). The defense-in-depth rationale (the `^-` anchor already filtered ordered items; the guard makes intent visible and future-proofs against regex broadening) is sound.

#2 (no opt-out input) — still open. `action.yml` has no `skip-misplaced-bump-check` (or equivalent) input. Consumers with pre-DT-145 hand-edited RELEASE.md history who bump to this action version will fail `check-release-log` with no clean bypass short of editing their file. Per the cascade infra convention, mechanical validators should default to shipping, not gating — an escape hatch input is needed before this merges broadly.

Cross-repo ([openfilter#104](https://github.com/PlainsightAI/openfilter/pull/104) `1d2af9fb`): the rewriter now normalizes bare `## Unreleased` and `## (unreleased)` headers to `## [Unreleased]` before splicing in the bullet, so every bump PR the rewriter produces will carry bracket-only Unreleased headings. This parser's strict `[Unreleased]`-only acceptance is now consistent with what the rewriter emits — the cross-repo contract is tight.

**Review by lucasmundim — COMMENTED** [review]
Thanks @stwilt — glad the cross-repo contract reads tight now.

On #2, pushed `2c8bd59` adding `skip-bump-bullet-checks` as an opt-in escape hatch. Design notes since this involved a real trade-off:

**Default stays strict.** Pushing back gently on "validators should default to shipping not gating" — `check-release-log` is itself a gate (it requires a RELEASE.md entry on `openfilter/**` changes), so I'm reading the org convention as "validators default to enforcement, with surgical opt-outs," not "validators default to advisory." Every misplaced-bump case we've seen in the wild has been a cascade bug, not a deliberate layout — making it skippable by default would mostly mean "let the bug slide" rather than "let intent through."

**But the escape hatch is real and exists,** because your reasoning on the rules-evolve angle stands: a future rule sharpening (or a time-critical hotfix sitting on top of legacy mis-placement) needs a clean way out. So the input is there, defaulting to `false`. When `true`:

- Both new rules are bypassed (intro + released-section). Nothing else relaxes — version-header parsing, ordering, and duplicate detection still run.
- `main.ts` emits a `core.warning()` so the bypass surfaces in the workflow's Annotations tab; reviewers see it even without reading the step log.
- README documents it under "Misplaced-bump validation → Escape hatch" with discouragement language framing it as a temporary safety valve, not a long-term opt-out. The framing is intentionally uncomfortable so fixing the underlying file remains the obvious next move.

Three new tests pin the behavior: intro-rule bypass works, released-section bypass works, and non-bump rules (duplicate version detection used as the canary) still fire regardless of the flag — so the input is a scalpel, not a blunderbuss.

66 tests pass (was 63, +3). Bundle ~unchanged.

**Review by shingonoide — APPROVED** [review]
The AST-driven rewrite is clean and the two validation rules are correctly scoped to top-level `list` tokens, so bump-shaped text inside fenced code blocks or HTML comments doesn't false-positive — that's a real win over the prior regex approach.

Verified end-to-end:
- Escape hatch `skip-bump-bullet-checks` defaults to `'false'` in `action.yml`, threads through `main.ts → Action.run → ChangelogParser.parseChangelog`, and the warning fires before parsing.
- `[Unreleased]` detection (`^\[\s*unreleased\s*\]$/i` → `entry.status === 'unreleased'`) correctly exempts the intended section from rule 2.
- `description` rebuilt via `token.raw` preserves per-line interior whitespace, and the regression pin test will catch future parser swaps that drop it.
- `marked` correctly declared as a runtime dependency, bundled into `build/dist/index.js`; the dist is in sync with `src` at HEAD.
- Test coverage is symmetric: FaceGuard intro pattern, SGSDA released pattern, both rule bypasses via escape hatch, escape hatch still enforces non-bump rules, and ordered-intro/non-bump-intro happy paths.

One cosmetic note (not a blocker): `validateNoBumpBulletsInRelease` omits the explicit `(t as Tokens.List).ordered` skip that its sibling `validateNoListInIntro` has. The `^-` anchor on `bumpBulletRegex` makes this safe today, but adding a matching `if ((t as Tokens.List).ordered) continue;` would make the symmetry obvious and future-proof against any broadening of the regex. Purely optional.

- `src/ChangelogParser.ts:185-202`

**Review by lucasmundim — COMMENTED** [review]
Thanks @shingonoide for the approval and the cross-check on the end-to-end plumbing.

Applied the cosmetic in `9557d26` — `validateNoBumpBulletsInRelease` now has the same `(t as Tokens.List).ordered` skip as its intro-validator sibling. The `^-` anchor was carrying the safety today; the explicit guard makes both rules' shape match and means a future broadening of `bumpBulletRegex` doesn't need to remember to also touch both call sites.

66 tests still pass; bundle ~unchanged. If the symmetry tweak dismissed the stale approval, happy to ping for a quick re-look — substantively nothing else moved.
